src/neo/Network/P2P/Peer.cs

```diff
@@ -205,28 +208,15 @@ private void OnStart(ChannelsConfig config)
         /// </summary>
         /// <param name="remote">The remote endpoint of TCP connection.</param>
         /// <param name="local">The local endpoint of TCP connection.</param>
-        private void OnTcpConnected(IPEndPoint remote, IPEndPoint local)
+        protected virtual void OnTcpConnected(IPEndPoint remote, IPEndPoint local)
         {
             ImmutableInterlocked.Update(ref ConnectingPeers, p => p.Remove(remote));
-            if (MaxConnections != -1 && ConnectedPeers.Count >= MaxConnections && !TrustedIpAddresses.Contains(remote.Address))
-            {
-                Sender.Tell(Tcp.Abort.Instance);
-                return;
-            }
```

Comment:
Can you create a method `Disconnect()` and move `Sender.Tell(Tcp.Abort.Instance);` to it? Then we can override it and send extra message.

Reply:
Done

Reply:
You didn't use it?

Reply:
I have adjusted the `Disconnect` arguments, as we need to pass the reason message.
```c#
protected void TcpDisconnect(byte[] lastMsg = null, bool connected = true)
protected void WsDisconnect(WebSocket ws, byte[] lastMsg = null)
```

Reply:
Maybe it can be:

```c#
protected virtual void TcpDisconnect(DisconnectReason reason);
```

Reply:
Then we can send messages in the overriding method.

Reply:
But, it's not a good place to put `DisconnectReason`  in `Peer`. Also, if we use the method above, we don't need override anymore, we just call the method in `LocalNode`.

Reply:
Because I want to move the checks to `Peer`.

Reply:
I may not get your point, If we move the checks to `Peer`, it seems there is no need to modify `LocalNode`, like overriding `OnTcpConnected`?

Reply:
Let's make a try.

Reply:
> Send messages in the overriding method

Got, done.